rpi_pico: Add hwinfo support

yonsch opened the pull request:
Adds hwinfo support for the rpi_pico series. Additionally, support for calling bootrom functions was added.

yonsch committed:
soc: rp2040: Enable bootrom functions

Support function and data lookup in the RPi's ROM.
The objects are obtained by including <pico/bootrom.h> and calling
rom_func_lookup or rom_data_lookup.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
drivers: hwinfo: Added unique ID support for RPi Pico

Added support for hwinfo's hwinfo_get_device_id for the
RPi Pico series.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch committed:
drivers: hwinfo: Added reset cause support for RPi Pico

Added support for hwinfo's reset cause functions, for the Raspberry Pi
Pico series.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

tejlmand reviewed (changes requested):
Generally looks good.

Minor request regarding compiler flag.

yonsch committed:
cmake: Add compiler property for disabling pointer arithmetic warnings

Add a compiler property for disabling pointer arithmetic warnings,
and implement that property for GCC.

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

yonsch commented on modules/hal_rpi_pico/CMakeLists.txt:
```diff
@@ -80,4 +80,21 @@ if(CONFIG_HAS_RPI_PICO)
+    ${rp2_common_dir}/hardware_flash/flash.c
+    PROPERTIES
+    COMPILE_FLAGS "-Wno-pointer-arith"
+  )
```
@tejlmand done

alexanderwachter commented on drivers/hwinfo/hwinfo_rpi_pico.c:
```diff
@@ -0,0 +1,65 @@
+{
+	/* The reset register can't be erased, nothing to do. */
+
+	return 0;
```
-ENOSUP would be more precise here.

yonsch committed:
boards: rpi_pico: Added HWINFO to supported features

Adds HWINFO to rpi_pico.yaml and documentation

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

alexanderwachter commented:
@yonsch I have to apologize. I told you the wrong errno.
-ENOSYS is the correct one. The API doc is not up to date.

yonsch committed:
hwinfo: Replace ENOTSUP with ENOSYS in documentation

Fix hwinfo documentation claiming API returns ENOTSUP for
unimplemented functions, with the correct ENOSYS

Signed-off-by: Yonatan Schachter <yonatan.schachter@gmail.com>

tejlmand reviewed (dismissed):
lgtm

alexanderwachter reviewed (dismissed):
Thanks for the doc change!

tejlmand reviewed (approved):
lgtm - re-approved
